### PR TITLE
Use c.Assert in integration-cli/docker_cli_logs_test.go

### PR DESCRIPTION
**Pull request by zhangjianfnst**
Part of #16756

Signed-off-by: Jian Zhang zhangjian.fnst@cn.fujitsu.com

### CODE REVIEW

File: integration-cli/docker_cli_logs_test.go [hunk 1]
```diff
@@ -321,9 +287,9 @@ func (s *DockerSuite) TestLogsFollowGoroutinesWithStdout(c *check.C) {
 	for {
 		select {
 		case <-t:
-			if n := getNGoroutines(); n > nroutines {
-				c.Fatalf("leaked goroutines: expected less than or equal to %d, got: %d", nroutines, n)
-			}
+			n := getNGoroutines()
+			c.Assert(n > nroutines, check.Equals, false, check.Commentf("leaked goroutines: expected less than or equal to %d, got: %d", nroutines, n))
```
Comment (HuKeping) [inline integration-cli/docker_cli_logs_test.go]: Maybe `c.Assert(n <= nroutines, check.Equals, true,...)` is better

File: integration-cli/docker_cli_logs_test.go [hunk 2]
```diff
@@ -232,17 +206,11 @@ func (s *DockerSuite) TestLogsSinceFutureFollow(c *check.C) {
 	since := now + 2
 	out, _ = dockerCmd(c, "logs", "-f", fmt.Sprintf("--since=%v", since), cleanedContainerID)
 	lines := strings.Split(strings.TrimSpace(out), "\n")
-	if len(lines) == 0 {
-		c.Fatal("got no log lines")
-	}
+	c.Assert(len(lines) == 0, check.Equals, false, check.Commentf("got no log lines"))
```
Comment (HuKeping) [inline integration-cli/docker_cli_logs_test.go]: adjust the logic to assert true rather than false might be better

**Comment by zhangjianfnst** [comment]
@HuKeping I agree with you, thanks.

File: integration-cli/docker_cli_logs_test.go [hunk 3]
```diff
@@ -169,9 +150,8 @@ func (s *DockerSuite) TestLogsFollowStopped(c *check.C) {
 	dockerCmd(c, "wait", cleanedContainerID)
 
 	logsCmd := exec.Command(dockerBinary, "logs", "-f", cleanedContainerID)
-	if err := logsCmd.Start(); err != nil {
-		c.Fatal(err)
-	}
+	err := logsCmd.Start()
+	c.Assert(err, checker.IsNil)
```
Comment (vdemeester) [inline integration-cli/docker_cli_logs_test.go]: You probably could do `c.Assert(logsCmd.Start(), checker.IsNil)`

**Comment by vdemeester** [comment]
@zhangjianfnst one nit, could use use `checker.…` for each checker ? (sometimes you use `check`, sometimes `checker` :stuck_out_tongue_closed_eyes:.

**Comment by zhangjianfnst** [comment]
@vdemeester Thanks for your comments.

File: integration-cli/docker_cli_logs_test.go [hunk 4]
```diff
@@ -19,13 +20,13 @@ func (s *DockerSuite) TestLogsContainerSmallerThanPage(c *check.C) {
...
 
+	cleanedContainerID := strings.TrimSpace(out)
 	dockerCmd(c, "wait", cleanedContainerID)
+
 	out, _ = dockerCmd(c, "logs", cleanedContainerID)
-	if len(out) != testLen+1 {
-		c.Fatalf("Expected log length of %d, received %d\n", testLen+1, len(out))
-	}
+
+	c.Assert(len(out), checker.Equals, testLen+1)
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: same result, but there does exist checker.HasLen to check length.

File: integration-cli/docker_cli_logs_test.go [hunk 5]
```diff
@@ -39,9 +40,7 @@ func (s *DockerSuite) TestLogsContainerBiggerThanPage(c *check.C) {
 
 	out, _ = dockerCmd(c, "logs", cleanedContainerID)
 
-	if len(out) != testLen+1 {
-		c.Fatalf("Expected log length of %d, received %d\n", testLen+1, len(out))
-	}
+	c.Assert(len(out), checker.Equals, testLen+1)
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: same comment
checker.HasLen

File: integration-cli/docker_cli_logs_test.go [hunk 6]
```diff
@@ -55,9 +54,7 @@ func (s *DockerSuite) TestLogsContainerMuchBiggerThanPage(c *check.C) {
...
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: same comment
checker.HasLen

File: integration-cli/docker_cli_logs_test.go [hunk 7]
```diff
@@ -72,21 +69,16 @@ func (s *DockerSuite) TestLogsTimestamps(c *check.C) {
 
 	lines := strings.Split(out, "\n")
 
-	if len(lines) != testLen+1 {
-		c.Fatalf("Expected log %d lines, received %d\n", testLen+1, len(lines))
-	}
+	c.Assert(len(lines), checker.Equals, testLen+1)
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: same comment
checker.HasLen

File: integration-cli/docker_cli_logs_test.go [hunk 8]
```diff
@@ -72,21 +69,16 @@ func (s *DockerSuite) TestLogsTimestamps(c *check.C) {
...
 	for _, l := range lines {
 		if l != "" {
 			_, err := time.Parse(timeutils.RFC3339NanoFixed+" ", ts.FindString(l))
-			if err != nil {
-				c.Fatalf("Failed to parse timestamp from %v: %v", l, err)
-			}
-			if l[29] != 'Z' { // ensure we have padded 0's
-				c.Fatalf("Timestamp isn't padded properly: %s", l)
-			}
+			c.Assert(err, checker.IsNil, check.Commentf("Failed to parse timestamp from %v: %v", l, err))
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: don't need to print err the second time. Assert will print it automatically.
```check.Commentf("Failed to parse timestamp from %v", l)```

File: integration-cli/docker_cli_logs_test.go [hunk 9]
```diff
@@ -72,21 +69,16 @@ func (s *DockerSuite) TestLogsTimestamps(c *check.C) {
...
+			// ensure we have padded 0's
+			c.Assert(l[29] == 'Z', checker.Equals, true, check.Commentf("Timestamp isn't padded properly: %s", l))
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: Maybe can simplify 
`c.Assert(l[29], checker.Equals, 'Z')`
Not sure if that will do the right thing or not.

File: integration-cli/docker_cli_logs_test.go [hunk 10]
```diff
@@ -142,23 +127,19 @@ func (s *DockerSuite) TestLogsTail(c *check.C) {
 
 	lines := strings.Split(out, "\n")
 
-	if len(lines) != 6 {
-		c.Fatalf("Expected log %d lines, received %d\n", 6, len(lines))
-	}
+	c.Assert(len(lines), checker.Equals, 6)
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: checker.HasLen

File: integration-cli/docker_cli_logs_test.go [hunk 11]
```diff
@@ -142,23 +127,19 @@ func (s *DockerSuite) TestLogsTail(c *check.C) {
...
+
 	out, _ = dockerCmd(c, "logs", "--tail", "all", cleanedContainerID)
 
 	lines = strings.Split(out, "\n")
 
-	if len(lines) != testLen+1 {
-		c.Fatalf("Expected log %d lines, received %d\n", testLen+1, len(lines))
-	}
+	c.Assert(len(lines), checker.Equals, testLen+1)
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: checker.HasLen

File: integration-cli/docker_cli_logs_test.go [hunk 12]
```diff
@@ -142,23 +127,19 @@ func (s *DockerSuite) TestLogsTail(c *check.C) {
...
+
 	out, _, _ = dockerCmdWithStdoutStderr(c, "logs", "--tail", "random", cleanedContainerID)
 
 	lines = strings.Split(out, "\n")
 
-	if len(lines) != testLen+1 {
-		c.Fatalf("Expected log %d lines, received %d\n", testLen+1, len(lines))
-	}
+	c.Assert(len(lines), checker.Equals, testLen+1)
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: checker.HasLen

File: integration-cli/docker_cli_logs_test.go [hunk 13]
```diff
@@ -194,16 +173,14 @@ func (s *DockerSuite) TestLogsSince(c *check.C) {
...
 	since := t + 1 // add 1s so log1 & log2 doesn't show up
 	out, _ = dockerCmd(c, "logs", "-t", fmt.Sprintf("--since=%v", since), name)
 
 	// Skip 2 seconds
 	unexpected := []string{"log1", "log2"}
 	for _, v := range unexpected {
-		if strings.Contains(out, v) {
-			c.Fatalf("unexpected log message returned=%v, since=%v\nout=%v", v, since, out)
-		}
+		c.Assert(out, checker.Not(checker.Contains), v, check.Commentf("unexpected log message returned=%v, since=%v\nout=%v", v, since, out))
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: shouldn't need to print `out` & `v`. Assert will print them automatically.

File: integration-cli/docker_cli_logs_test.go [hunk 14]
```diff
@@ -212,13 +189,9 @@ func (s *DockerSuite) TestLogsSince(c *check.C) {
 		exec.Command(dockerBinary, "logs", "-t", "--since=0", name),
 	} {
 		out, _, err = runCommandWithOutput(cmd)
-		if err != nil {
-			c.Fatalf("failed to log container: %s, %v", out, err)
-		}
+		c.Assert(err, checker.IsNil, check.Commentf("failed to log container: %s, %v", out, err))
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: don't need to print `err` in commentf.

File: integration-cli/docker_cli_logs_test.go [hunk 15]
```diff
@@ -212,13 +189,9 @@ func (s *DockerSuite) TestLogsSince(c *check.C) {
...
 		for _, v := range expected {
-			if !strings.Contains(out, v) {
-				c.Fatalf("'%v' does not contain=%v\nout=%s", cmd.Args, v, out)
-			}
+			c.Assert(out, checker.Contains, v, check.Commentf("'%v' does not contain=%v\nout=%s", cmd.Args, v, out))
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: don't need to print `out` & `v` in commentf.

File: integration-cli/docker_cli_logs_test.go [hunk 16]
```diff
@@ -232,17 +205,11 @@ func (s *DockerSuite) TestLogsSinceFutureFollow(c *check.C) {
 	since := now + 2
 	out, _ = dockerCmd(c, "logs", "-f", fmt.Sprintf("--since=%v", since), cleanedContainerID)
 	lines := strings.Split(strings.TrimSpace(out), "\n")
-	if len(lines) == 0 {
-		c.Fatal("got no log lines")
-	}
+	c.Assert(len(lines) != 0, checker.Equals, true, check.Commentf("got no log lines"))
```
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: commentf can be normal `//` comment on same line or previous line.

File: integration-cli/docker_cli_logs_test.go (hunk 9 above)
Comment (zhangjianfnst) [inline integration-cli/docker_cli_logs_test.go]: @MHBauer I tried that, but FAIL occurred.

    ----------------------------------------------------------------------
    FAIL: docker_cli_logs_test.go:60: DockerSuite.TestLogsTimestamps
    
    docker_cli_logs_test.go:82:
        // ensure we have padded 0's
        c.Assert(l[29], checker.Equals, 'Z')
    ... obtained uint8 = 0x5a
    ... expected int32 = 90

File: integration-cli/docker_cli_logs_test.go [hunk 17]
```diff
@@ -72,21 +69,16 @@ func (s *DockerSuite) TestLogsTimestamps(c *check.C) {
...
 			_, err := time.Parse(timeutils.RFC3339NanoFixed+" ", ts.FindString(l))
-			if err != nil {
-				c.Fatalf("Failed to parse timestamp from %v: %v", l, err)
-			}
-			if l[29] != 'Z' { // ensure we have padded 0's
-				c.Fatalf("Timestamp isn't padded properly: %s", l)
-			}
+			c.Assert(err, checker.IsNil, check.Commentf("Failed to parse timestamp from %v", l))
+			// ensure we have padded 0's
+			c.Assert(l[29] == 'Z', checker.Equals, true, check.Commentf("Timestamp isn't padded properly: %s", l))
```
Comment (cpuguy83) [inline integration-cli/docker_cli_logs_test.go]: `c.Assert(l[29], checker.Equals, 'Z')`

File: integration-cli/docker_cli_logs_test.go [hunk 18]
```diff
@@ -19,13 +20,13 @@ func (s *DockerSuite) TestLogsContainerSmallerThanPage(c *check.C) {
 	testRequires(c, DaemonIsLinux)
 	testLen := 32767
 	out, _ := dockerCmd(c, "run", "-d", "busybox", "sh", "-c", fmt.Sprintf("for i in $(seq 1 %d); do echo -n =; done; echo", testLen))
-	cleanedContainerID := strings.TrimSpace(out)
 
+	cleanedContainerID := strings.TrimSpace(out)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_logs_test.go]: can we change this to `id`?

File: integration-cli/docker_cli_logs_test.go (hunk 17 above)
Comment (MHBauer) [inline integration-cli/docker_cli_logs_test.go]: Apparently it doesn't work. https://github.com/docker/docker/pull/16969#discussion_r41909549

Comment (zhangjianfnst) [inline integration-cli/docker_cli_logs_test.go]: Casting type works.
`c.Assert(l[29], checker.Equals, uint8('Z'))`

File: integration-cli/docker_cli_logs_test.go (hunk 18 above)
Comment (zhangjianfnst) [inline integration-cli/docker_cli_logs_test.go]: Yes, `id` is much briefer.

**Comment by vdemeester** [comment]
LGTM 🐼

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**